### PR TITLE
Feature/layer dependency checker

### PR DESCRIPTION
**Pull request by juanmatias**
## What?
* Create a task to check layer dependencies
* Example of possible command: 
```shell
# for a full output
❯ leverage run layer_dependency
# for a summarized output
❯ leverage run layer_dependency\['summary=True'\]
```

## Why?
* When about to apply a layer it is desirable to check what layers the current one depends on

## References
* In the first iteration the check can be done based on remote states. I.e. the layers owning the remote states should be applied already.
* Then other data resources can be included in the checking process.

### CODE REVIEW

File: build.py [hunk 1]
```diff
@@ -23,4 +28,98 @@ def decrypt():
...
+      #   config = {
+      #     region  = var.region
+      #     profile = var.profile
+      #     bucket  = var.bucket
+      #     key     = "apps-prd/k8s-eks/network/terraform.tfstate"
+      #   }
+      # }
+      remote_state_resource_name = 'terraform_remote_state'
+      for data in config_file_contents:
+              if 'data' in data.keys():
```
Comment (Franr) [inline build.py:95]: you could flatten all this structure by skipping the current value if it doesn't met the criteria

```suggestion
              if not 'data' in data.keys():
                  continue
```

the same for the rest of the `ifs` inside the loop

File: build.py [hunk 2]
```diff
@@ -23,4 +28,98 @@ def decrypt():
...
+    if summary.lower() == 'true':
+        deps = {'this': []}
+        for rs in remote_states:
+            deps['this'].append(remote_states[rs]['key'])
+        remote_states = deps
+
+    try:
+        if quiet == 'False': print("Note layer dependency is calculated using remote states.\nNevertheless, other sort of dependencies could exist without this kind of resources,\ne.g. if you relay on some resource created in a different layer and not referenced here.")
+        if len(remote_states) > 1:
+          if len(remote_states) > 2:
```
Comment (Franr) [inline build.py:50]: 2 space indentation? :crying_cat_face:

File: build.py [hunk 3]
```diff
@@ -23,4 +28,98 @@ def decrypt():
...
+def layer_dependency(summary='False',quiet='False'):
+    """Check layer dependency from the current layers"""
+    try:
+        remote_states = _layer_dependency(path.get_working_path())
+    except Exception as e:
+        print('Error getting deps:',e)
+
+    if summary.lower() == 'true':
+        deps = {'this': []}
+        for rs in remote_states:
```
Comment (Franr) [inline build.py:43]: if there was an exception (line 39), `remote_states` won't be defined so this will break

File: build.py [hunk 4]
```diff
@@ -5,6 +5,11 @@
 from leverage import task
 from leverage import path
 
+import hcl2
+import json
```
Comment (Franr) [inline build.py:9]: standard libraries goes first, then 3rd party and then locals

https://peps.python.org/pep-0008/#imports

File: build.py [hunk 5]
```diff
@@ -23,4 +28,98 @@ def decrypt():
...
+      # #########################################
+      # Where data is being used
+      # #########################################
+      remote_state_usage_regex = 'data\.terraform_remote_state\.([a-z_\-0-9]+)\..+'
+      files = glob.glob(f"{directory}/*tf")
+      for this_file in files:
+          with open(this_file, 'r') as f:
+              this_file_data = f.readlines()
+              for line in this_file_data:
+                  match_object = re.search(remote_state_usage_regex, line)
```
Comment (Franr) [inline build.py:119]:
```suggestion
                  if match_object := re.search(remote_state_usage_regex, line)
```
You could use the walrus operator here :slightly_smiling_face:

File: build.py (hunk 1 above)
Comment (juanmatias) [reply build.py]: great suggestion!

File: build.py (hunk 2 above)
Comment (juanmatias) [reply build.py]: yeah, well... my IDE decided to do that... fixed... 😄

File: build.py (hunk 5 above)
Comment (juanmatias) [reply build.py]: also a great suggestion!

File: build.py (hunk 4 above)
Comment (juanmatias) [reply build.py]: nice

File: build.py (hunk 3 above)
Comment (juanmatias) [reply build.py:44]: fixed

File: build.py [hunk 6]
```diff
@@ -23,4 +31,177 @@ def decrypt():
...
+        sys.exit(0)
+
+    if summary.lower() == 'true':
+        deps = {'this': []}
+        for rs in remote_states:
+            deps['this'].append(remote_states[rs]['key'])
+        remote_states = deps
+
+    try:
+        if quiet == 'False': print("Note layer dependency is calculated using remote states.\nNevertheless, other sort of dependencies could exist without this kind of resources,\ne.g. if you relay on some resource created in a different layer and not referenced here.")
```
Comment (angelofenoglio) [inline build.py:52]:
```suggestion
        if quiet == 'False': print("Note layer dependency is calculated using remote states.\nNevertheless, other sort of dependencies could exist without this kind of resources,\ne.g. if you rely on some resource created in a different layer and not referenced here.")
```

File: build.py [hunk 7]
```diff
@@ -23,4 +31,177 @@ def decrypt():
...
+        for this_file in files:
+            with open(this_file, 'r') as f:
+                this_file_data = f.readlines()
+                for line in this_file_data:
+                    if match_object := re.search(remote_state_usage_regex, line):
+                        if not match_object.group(1) in remote_state_lines.keys():
+                            remote_state_lines[match_object.group(1)] = [this_file]
+                        else:
+                            if not this_file in remote_state_lines[match_object.group(1)]:
+                                remote_state_lines[match_object.group(1)].append(this_file)
```
Comment (angelofenoglio) [inline build.py:97]: A `defaultdict(list)` would be the ideal structure to use in here, simply appending blindly like in the last line for every match would take care of this logic

File: build_deplayerchk.py [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+    Load local terraform values from all *tf files
+    """
+
+    locals_file_contents = {}
+
+    files = glob.glob(f"./*tf")
+
+    return _load_hclfile(files,_load_locals_aux)
+
+    #for this_file in files:
```
Comment (Franr) [inline build_deplayerchk.py:255]: :warning:

**Review by exequielrafaela — DISMISSED** [review]

File: build_deplayerchk.py (hunk 8 above)
Comment (juanmatias) [reply build_deplayerchk.py]: 👍🏾

**Review by Franr — APPROVED** [review]

**Review by Franr — APPROVED** [review]

**Pull request merged**